scripts/check-node-version-alignment.js

```diff
@@ -0,0 +1,79 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { readFile, readdir } from "node:fs/promises";
+import { resolve, dirname } from "node:path";
+import { fileURLToPath } from "node:url";
+
+/**
+ * This script checks whether every place we set a Node version is set to the same version.
+ *
+ * I've opted to avoid adding e.g. TOML and YAML parsers, and just assume
+ * specific line formatting, to write the simplest thing that works in the least
+ * amount of time.
+ *
+ * It is run in `.github/workflows/lint.yaml`.
+ */
+
+const __dirname = dirname(fileURLToPath(import.meta.url));
+const load = (path) => readFile(resolve(__dirname, "../", path), "utf-8");
+
+const packageJson = JSON.parse(await load("package.json"));
+const lockFile = JSON.parse(await load("package-lock.json"));
+const dockerFileNodeVersion = (await load("Dockerfile"))
+  .split("\n")
+  .find((line) => line.startsWith("FROM node:"))
+  .split("FROM node:")[1]
+  .split("-alpine")[0];
+// Yes, we're assuming `netlify.toml` to contain a line in exactly this format:
+//   environment = { NODE_VERSION = "20.12.0", NPM_VERSION = "10.2.4" }
+const netlifyNodeVersion = (await load("netlify.toml"))
+  .split("\n")
+  .find(line => line.includes('NODE_VERSION = "'))
+  .split('NODE_VERSION = "')[1]
+  .split('", NPM_VERSION')[0];
+
+const circleNodeVersions = (await load(".circleci/config.yml"))
+  .split("\n")
+  .filter(line => line.includes("cimg/node:"))
+  .map(line => line.split("cimg/node:")[1]);
+
+const esbuildVersion = (await load("esbuild.cronjobs.js"))
+  .split("\n")
+  .find(line => line.includes("target:"))
+  .split('"node')[1]
+  .split('",')[0];
+
+const ghaWorkflowPaths = await readdir(resolve(__dirname, "../.github/workflows"));
+const ghaWorkflows = await Promise.all(
+  ghaWorkflowPaths.map(async (workflowFileName) =>
+    [workflowFileName, await load(`.github/workflows/${workflowFileName}`)],
+  ),
+);
+const ghaVersions = ghaWorkflows.flatMap(([filename, contents]) => {
+  return contents.split("\n")
+    .filter(line => line.includes("node-version:"))
+    .map(line => [filename, line.split("node-version: ")[1].replaceAll("'", "")]);
+});
+
+const getMinorOnly = (version) => {
+  const versionParts = version.split(".");
+  return `${versionParts[0]}.${versionParts[1]}`;
+};
+const checkVersion = (source, version) => {
+  const voltaVersion = packageJson.volta.node;
+  if (getMinorOnly(voltaVersion) !== getMinorOnly(version)) {
+    console.error(`Error: version [${version}] in ${source} does not match the Volta version ([${voltaVersion}]).`)
+    process.exit(1);
+  }
+};
+
+checkVersion("the package.json engines", packageJson.engines.node);
+checkVersion("the package-lock.json engines", lockFile.packages[""].engines.node);
```

Comment:
Did you mean to check the `[...].engines.npm` version here too? I noticed that it's checked above in `.netlify.toml`

Reply:
Currently the script is only checking whether the Node versions are aligned everywhere, as I'm not really sure what to do with the NPM versions. I think there's something to be said for not specifying it and going with the version bundled with whatever Node version we're using, but there's also something to be said for disconnecting the two and choosing an explicit npm version ourselves (although I don't think Dependabot can check any of the places we can set the npm version). What do you think?

Reply:
I don't know how much of a problem it is, but I do see complaints in our CI logs if the NPM version doesn't match. I don't think it's urgent though we can add it later, just having the script to check npm version will be a huge improvement, thanks for doing that.